src/uploaders.rs

```diff
@@ -127,134 +121,76 @@ impl Uploader {
                         response
                     )));
                 }
-                let readme_path = if let Some(rendered) = readme {
-                    let mut handle = req.app().handle();
-                    let readme_path =
-                        format!("/{}", Uploader::readme_path(&krate.name, &vers.to_string()));
-                    let response = {
-                        let mut response = Vec::new();
-                        {
-                            let readme_len = rendered.len();
-                            let mut cursor = io::Cursor::new(rendered.into_bytes());
-                            let mut s3req = bucket.put(
-                                &mut handle,
-                                &readme_path,
-                                &mut cursor,
-                                "text/html",
-                                readme_len as u64,
-                            );
-                            s3req
-                                .write_function(|data| {
-                                    response.extend(data);
-                                    Ok(data.len())
-                                })
-                                .unwrap();
-                            s3req.perform().chain_error(|| {
-                                internal(&format_args!(
-                                    "failed to upload readme to S3: `{}`",
-                                    readme_path
-                                ))
-                            })?;
-                        }
-                        response
-                    };
-                    if handle.response_code().unwrap() != 200 {
-                        if let Err(e) = self.delete(req.app().clone(), &crate_path) {
```

Comment:
This `delete` call is here to remove the uploaded crate if the readme couldn't be uploaded. It technically wasn't removed. In the new code, I'm creating the `Bomb` for the uploaded crate before uploading the readme so that if the readme upload fails a `CargoResult::Error()` is returned, triggering the Bomb and deleting the uploaded crate file. I hope I got the `Bomb` behavior correct.